src/t_set.c

```diff
@@ -122,17 +122,16 @@ int setTypeAddAux(robj *set, char *str, size_t len, int64_t llval, int str_is_sd
         /* Avoid duping the string if it is an sds string. */
         sds sdsval = str_is_sds ? (sds)str : sdsnewlen(str, len);
         dict *ht = set->ptr;
-        dictEntry *de = dictAddRaw(ht,sdsval,NULL);
-        if (de && sdsval == str) {
-            /* String was added but we don't own this sds string. Dup it and
-             * replace it in the dict entry. */
-            dictSetKey(ht,de,sdsdup((sds)str));
-            dictSetVal(ht,de,NULL);
-        } else if (!de && sdsval != str) {
-            /* String was already a member. Free our temporary sds copy. */
+        long index = dictKeyIndex(ht, sdsval, dictHashKey(ht, sdsval), NULL);
+        if (index >= 0) {
+            /* Key doesn't already exist in the set. Add it but dup the key. */
+            if (sdsval == str) sdsval = sdsdup(sdsval);
+            dictInsertAtIndex(ht, sdsval, index);
+        } else if (sdsval != str) {
+            /* String is already a member. Free our temporary sds copy. */
```

Comment:
> i understand the reason is that we want to avoid excessive sdsdup and sdsfree if the element is already there, right?

Exactly. A kind of "two-phase add"...

Are you proposing we can essentially keep this dictKeyIndex + dictInsertAtIndex if we just hide hash (easy, just add a wrapper to dictKeyIndex that computes the hash internally) and hide that we're working with an index directly into the table? For example make the index an opaque type or returning a `dictEntry**` bucket-pointer instead of an index?

> or maybe it's enough to return a `dictEntry*` and if we see the LSB set, we realize that it's actually the hash table itself (and then dictSetKey can still update it)?

If we return a dictEntry-pointer which is a key, there is no way to replace it other than to hash the key again to find the bucket. That's why dictSetKey doesn't work well in this case.

Reply:
I chose to use a bucket pointer (`dictEntry **`) instead of index. I replaced the functions with `dictFindBucketForInsert` and `dictInsertIntoBucket`. The key hashing and the index is hidden within the functions. 

(I didn't use an opaque struct `dictEntryRef` for the intermediate values, because the user can't allocate an opaque value on the stack, so it'd have to be allocated.)